`instrumentCode` task fails with EAP version of Android Studio

https://github.com/JetBrains/gradle-intellij-plugin/pull/883 tried to fix a problem with proper value of `org.jetbrains.intellij.tasks.IntelliJInstrumentCodeTask#getCompilerVersion`. And looks like it works as expected for release version of Android Studio.

But if you use pre-release version of Android Studio (like we do in [educational-plugin](https://github.com/JetBrains/educational-plugin)), `instrumentCode` task still fails.
As I can see, it happens because of `-EAP-SNAPSHOT` suffix added to `compilerVersion` by gradle-intellij-plugin. If you manually set the same value but without this suffix, it will work

You can reproduce it with [this](https://github.com/JetBrains/educational-plugin/commit/cdae20938c0b16d392f72d0fccd0fcc5d2181866) commit

<details><summary>stacktrace</summary>

```
[gradle-intellij-plugin :educational-coreeducational-core:educational-core:instrumentCode] Error when resolving dependency: DefaultExternalModuleDependency{group='com.jetbrains.intellij.java', name='java-compiler-ant-tasks', version='212.5457.46-EAP-SNAPSHOT', configuration='default'}
org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$ArtifactResolveException: Could not resolve all files for configuration ':educational-core:detachedConfiguration3'.
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.rethrowFailure(DefaultConfiguration.java:1503)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$3700(DefaultConfiguration.java:159)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$DefaultResolutionHost.rethrowFailure(DefaultConfiguration.java:2139)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationFileCollection.visitContents(DefaultConfiguration.java:1475)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.visitContents(DefaultConfiguration.java:509)
	at org.gradle.api.internal.file.AbstractFileCollection.getFiles(AbstractFileCollection.java:127)
	at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration_Decorated.getFiles(Unknown Source)
	at org.jetbrains.intellij.utils.DependenciesDownloader.downloadFromMultipleRepositories(DependenciesDownloader.kt:36)
	at org.jetbrains.intellij.IntelliJPlugin$configureInstrumentation$$inlined$forEach$lambda$1$8.call(IntelliJPlugin.kt:735)
	at org.jetbrains.intellij.IntelliJPlugin$configureInstrumentation$$inlined$forEach$lambda$1$8.call(IntelliJPlugin.kt:67)
	at org.gradle.api.internal.provider.DefaultProvider.calculateOwnValue(DefaultProvider.java:72)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:103)
	at org.gradle.api.internal.provider.Collectors$ElementsFromCollectionProvider.collectEntries(Collectors.java:216)
	at org.gradle.api.internal.provider.AbstractCollectionProperty$CollectingSupplier.calculateValue(AbstractCollectionProperty.java:337)
	at org.gradle.api.internal.provider.AbstractCollectionProperty.finalValue(AbstractCollectionProperty.java:189)
	at org.gradle.api.internal.provider.AbstractCollectionProperty.finalValue(AbstractCollectionProperty.java:37)
	at org.gradle.api.internal.provider.AbstractProperty.finalizeNow(AbstractProperty.java:236)
	at org.gradle.api.internal.provider.AbstractProperty.beforeRead(AbstractProperty.java:230)
	at org.gradle.api.internal.provider.AbstractProperty.calculateOwnValue(AbstractProperty.java:126)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.getOrNull(AbstractMinimalProvider.java:93)
	at org.gradle.api.internal.provider.ProviderResolutionStrategy$1.resolve(ProviderResolutionStrategy.java:27)
	at org.gradle.util.internal.DeferredUtil.unpack(DeferredUtil.java:59)
	at org.gradle.util.internal.DeferredUtil.unpackOrNull(DeferredUtil.java:49)
	at org.gradle.api.internal.tasks.properties.InputParameterUtils.prepareInputParameterValue(InputParameterUtils.java:39)
	at org.gradle.api.internal.tasks.properties.InputParameterUtils.prepareInputParameterValue(InputParameterUtils.java:31)
	at org.gradle.api.internal.tasks.execution.TaskExecution.lambda$visitRegularInputs$1(TaskExecution.java:315)
	at org.gradle.internal.execution.fingerprint.impl.DefaultInputFingerprinter$InputCollectingVisitor.visitInputProperty(DefaultInputFingerprinter.java:106)
	at org.gradle.api.internal.tasks.execution.TaskExecution.visitRegularInputs(TaskExecution.java:315)
	at org.gradle.internal.execution.fingerprint.impl.DefaultInputFingerprinter.fingerprintInputProperties(DefaultInputFingerprinter.java:61)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.captureExecutionStateWithOutputs(CaptureStateBeforeExecutionStep.java:193)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.lambda$captureExecutionState$1(CaptureStateBeforeExecutionStep.java:141)
	at org.gradle.internal.execution.steps.BuildOperationStep$1.call(BuildOperationStep.java:37)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.internal.execution.steps.BuildOperationStep.operation(BuildOperationStep.java:34)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.captureExecutionState(CaptureStateBeforeExecutionStep.java:130)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.lambda$execute$0(CaptureStateBeforeExecutionStep.java:75)
	at java.base@11.0.8/java.util.Optional.flatMap(Optional.java:294)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:75)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:50)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.executeWithNoEmptySources(SkipEmptyWorkStep.java:249)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:86)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:54)
	at org.gradle.internal.execution.steps.RemoveUntrackedExecutionStateStep.execute(RemoveUntrackedExecutionStateStep.java:32)
	at org.gradle.internal.execution.steps.RemoveUntrackedExecutionStateStep.execute(RemoveUntrackedExecutionStateStep.java:21)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
	at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:43)
	at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:31)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.lambda$execute$0(AssignWorkspaceStep.java:40)
	at org.gradle.api.internal.tasks.execution.TaskExecution$4.withWorkspace(TaskExecution.java:287)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:40)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:30)
	at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:37)
	at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:27)
	at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:44)
	at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:33)
	at org.gradle.internal.execution.impl.DefaultExecutionEngine$1.execute(DefaultExecutionEngine.java:76)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:144)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:133)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:333)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:320)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:313)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:299)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:143)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:227)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:218)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:140)
	at org.gradle.execution.plan.DefaultPlanExecutor.process(DefaultPlanExecutor.java:72)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.executeWithServices(DefaultTaskExecutionGraph.java:144)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.execute(DefaultTaskExecutionGraph.java:129)
	at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:42)
	at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:51)
	at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor$ExecuteTasks.call(BuildOperationFiringBuildWorkerExecutor.java:54)
	at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor$ExecuteTasks.call(BuildOperationFiringBuildWorkerExecutor.java:43)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor.execute(BuildOperationFiringBuildWorkerExecutor.java:40)
	at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$executeTasks$7(DefaultBuildLifecycleController.java:165)
	at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:238)
	at org.gradle.internal.model.StateTransitionController.lambda$tryTransition$8(StateTransitionController.java:174)
	at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:44)
	at org.gradle.internal.model.StateTransitionController.tryTransition(StateTransitionController.java:174)
	at org.gradle.internal.build.DefaultBuildLifecycleController.executeTasks(DefaultBuildLifecycleController.java:165)
	at org.gradle.internal.build.DefaultBuildWorkGraphController$DefaultBuildWorkGraph.runWork(DefaultBuildWorkGraphController.java:142)
	at org.gradle.composite.internal.DefaultBuildController.doBuild(DefaultBuildController.java:231)
	at org.gradle.internal.Factories$1.create(Factories.java:31)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:270)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:119)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:124)
	at org.gradle.composite.internal.DefaultBuildController.doRun(DefaultBuildController.java:204)
	at org.gradle.composite.internal.DefaultBuildController.access$000(DefaultBuildController.java:51)
	at org.gradle.composite.internal.DefaultBuildController$BuildOpRunnable.run(DefaultBuildController.java:264)
	at java.base@11.0.8/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base@11.0.8/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base@11.0.8/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base@11.0.8/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base@11.0.8/java.lang.Thread.run(Thread.java:834)
Caused by: org.gradle.internal.resolve.ModuleVersionNotFoundException: Could not find com.jetbrains.intellij.java:java-compiler-ant-tasks:212.5457.46-EAP-SNAPSHOT.
Searched in the following locations:
  - https://repo.maven.apache.org/maven2/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
  - https://repo.maven.apache.org/maven2/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
  - https://cache-redirector.jetbrains.com/intellij-dependencies/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
  - https://cache-redirector.jetbrains.com/intellij-dependencies/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
  - https://maven.pkg.jetbrains.space/kotlin/p/kotlin/kotlin-js-wrappers/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
  - https://maven.pkg.jetbrains.space/kotlin/p/kotlin/kotlin-js-wrappers/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
  - file:/Users/Arseniy.Pendryak/Programming/educational-plugin/dependencies/studio-2021.2.1.8/Android Studio Preview.app/Contents/com.jetbrains.intellij.java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/ivy-212.5457.46-EAP-SNAPSHOT.xml
  - file:/Users/Arseniy.Pendryak/Programming/educational-plugin/dependencies/studio-2021.2.1.8/Android Studio Preview.app/Contents/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT-2-withKotlin.xml
  - https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/snapshots/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
  - https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/snapshots/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
Required by:
    project :educational-core

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':educational-core:instrumentCode'.
> Error while evaluating property 'compilerClassPathFromMaven' of task ':educational-core:instrumentCode'
   > Failed to calculate the value of task ':educational-core:instrumentCode' property 'compilerClassPathFromMaven'.
      > Could not resolve all files for configuration ':educational-core:detachedConfiguration3'.
         > Could not find com.jetbrains.intellij.java:java-compiler-ant-tasks:212.5457.46-EAP-SNAPSHOT.
           Searched in the following locations:
             - https://repo.maven.apache.org/maven2/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
             - https://repo.maven.apache.org/maven2/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
             - https://cache-redirector.jetbrains.com/intellij-dependencies/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
             - https://cache-redirector.jetbrains.com/intellij-dependencies/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
             - https://maven.pkg.jetbrains.space/kotlin/p/kotlin/kotlin-js-wrappers/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
             - https://maven.pkg.jetbrains.space/kotlin/p/kotlin/kotlin-js-wrappers/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
             - file:/Users/Arseniy.Pendryak/Programming/educational-plugin/dependencies/studio-2021.2.1.8/Android Studio Preview.app/Contents/com.jetbrains.intellij.java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/ivy-212.5457.46-EAP-SNAPSHOT.xml
             - file:/Users/Arseniy.Pendryak/Programming/educational-plugin/dependencies/studio-2021.2.1.8/Android Studio Preview.app/Contents/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT-2-withKotlin.xml
             - https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/snapshots/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/maven-metadata.xml
             - https://cache-redirector.jetbrains.com/www.jetbrains.com/intellij-repository/snapshots/com/jetbrains/intellij/java/java-compiler-ant-tasks/212.5457.46-EAP-SNAPSHOT/java-compiler-ant-tasks-212.5457.46-EAP-SNAPSHOT.pom
           Required by:
               project :educational-core
```

</details>

Related to #884

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,9 @@
 # Changelog
 
 ## next
+- Fix compiler resolution for EAP versions [#906](../../issues/906)
+
+## 1.4.0
 - Fixed JBR resolving for MacOSX M1
 - Fixed compiler resolution for long build numbers [#883](../../issues/883)
 - Build number fallback when `product-info.json` is missing [#880](../../issues/880)

--- a/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
@@ -32,6 +32,8 @@ import org.gradle.tooling.BuildException
 import org.jetbrains.gradle.ext.IdeaExtPlugin
 import org.jetbrains.gradle.ext.ProjectSettings
 import org.jetbrains.gradle.ext.TaskTriggersConfig
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_SUFFIX_EAP_CANDIDATE
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_SUFFIX_SNAPSHOT
 import org.jetbrains.intellij.dependency.IdeaDependency
 import org.jetbrains.intellij.dependency.IdeaDependencyManager
 import org.jetbrains.intellij.dependency.PluginDependency
@@ -700,11 +702,11 @@ open class IntelliJPlugin : Plugin<Project> {
                         val localPath = extension.localPath.orNull
                         val ideaDependency = setupDependenciesTask.idea.get()
 
-                        if (localPath.isNullOrBlank() && version.endsWith("-SNAPSHOT")) {
+                        if (localPath.isNullOrBlank() && version.endsWith(RELEASE_SUFFIX_SNAPSHOT)) {
                             val type = extension.getVersionType()
                             if (version == IntelliJPluginConstants.DEFAULT_IDEA_VERSION && listOf("CL", "RD", "PY").contains(type)) {
                                 ideProductInfo(ideaDependency.classes)?.buildNumber?.let { buildNumber ->
-                                    Version.parse(buildNumber).let { v -> "${v.major}.${v.minor}-EAP-CANDIDATE-SNAPSHOT" }
+                                    Version.parse(buildNumber).let { v -> "${v.major}.${v.minor}$RELEASE_SUFFIX_EAP_CANDIDATE" }
                                 } ?: version
                             } else {
                                 when (type) {
@@ -716,7 +718,7 @@ open class IntelliJPlugin : Plugin<Project> {
                             }
                         } else {
                             val isEap = localPath?.let { ideProductInfo(ideaDependency.classes)?.versionSuffix == "EAP" } ?: false
-                            val eapSuffix = IntelliJPluginConstants.EAP_SUFFIX.takeIf { isEap } ?: ""
+                            val eapSuffix = IntelliJPluginConstants.RELEASE_SUFFIX_EAP.takeIf { isEap } ?: ""
 
                             IdeVersion.createIdeVersion(ideaDependency.buildNumber)
                                 .stripExcessComponents()
@@ -755,9 +757,9 @@ open class IntelliJPlugin : Plugin<Project> {
                                  * for it fails - not all versions have a corresponding -EAP-SNAPSHOT version present
                                  * in the snapshot repository.
                                  */
-                                if (compilerVersion.endsWith(IntelliJPluginConstants.EAP_SUFFIX)) {
+                                if (compilerVersion.endsWith(IntelliJPluginConstants.RELEASE_SUFFIX_EAP)) {
                                     val nonEapVersion = compilerVersion.replace(
-                                        IntelliJPluginConstants.EAP_SUFFIX, ""
+                                        IntelliJPluginConstants.RELEASE_SUFFIX_EAP, ""
                                     )
                                     downloadCompiler(nonEapVersion)
                                 } else {

--- a/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
@@ -368,8 +368,8 @@ open class IntelliJPlugin : Plugin<Project> {
                         else -> setupDependenciesTask.idea.get().classes.let { ideaClasses ->
                             ideProductInfo(ideaClasses)
                                 ?.run { "$productCode-$version" }
-                                // Fall back on build number if product-info.json is not present, this is the case
-                                // for recent versions of Android Studio.
+                            // Fall back on build number if product-info.json is not present, this is the case
+                            // for recent versions of Android Studio.
                                 ?: ideBuildNumber(ideaClasses)
                         }
                     },
@@ -734,7 +734,7 @@ open class IntelliJPlugin : Plugin<Project> {
                         if (compilerVersion == IntelliJPluginConstants.DEFAULT_IDEA_VERSION ||
                             Version.parse(compilerVersion) >= Version(183, 3795, 13)
                         ) {
-                            val downloadCompiler: (String) -> List<File> = { version ->
+                            val downloadCompiler = { version: String ->
                                 dependenciesDownloader.downloadFromMultipleRepositories(logCategory(), {
                                     create(
                                         group = "com.jetbrains.intellij.java",

--- a/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
@@ -716,7 +716,7 @@ open class IntelliJPlugin : Plugin<Project> {
                             }
                         } else {
                             val isEap = localPath?.let { ideProductInfo(ideaDependency.classes)?.versionSuffix == "EAP" } ?: false
-                            val eapSuffix = "-EAP-SNAPSHOT".takeIf { isEap } ?: ""
+                            val eapSuffix = IntelliJPluginConstants.EAP_SUFFIX.takeIf { isEap } ?: ""
 
                             IdeVersion.createIdeVersion(ideaDependency.buildNumber)
                                 .stripExcessComponents()
@@ -732,17 +732,37 @@ open class IntelliJPlugin : Plugin<Project> {
                         if (compilerVersion == IntelliJPluginConstants.DEFAULT_IDEA_VERSION ||
                             Version.parse(compilerVersion) >= Version(183, 3795, 13)
                         ) {
-                            dependenciesDownloader.downloadFromMultipleRepositories(logCategory(), {
-                                create(
-                                    group = "com.jetbrains.intellij.java",
-                                    name = "java-compiler-ant-tasks",
-                                    version = compilerVersion,
-                                )
-                            }, {
-                                listOf(
-                                    "${extension.intellijRepository.get()}/${releaseType(compilerVersion)}",
-                                    IntelliJPluginConstants.INTELLIJ_DEPENDENCIES,
-                                ).map { url -> mavenRepository(url) }
+                            val downloadCompiler: (String) -> List<File> = { version ->
+                                dependenciesDownloader.downloadFromMultipleRepositories(logCategory(), {
+                                    create(
+                                        group = "com.jetbrains.intellij.java",
+                                        name = "java-compiler-ant-tasks",
+                                        version = version,
+                                    )
+                                }, {
+                                    listOf(
+                                        "${extension.intellijRepository.get()}/${releaseType(version)}",
+                                        IntelliJPluginConstants.INTELLIJ_DEPENDENCIES
+                                    ).map(::mavenRepository)
+                                })
+                            }
+
+                            runCatching {
+                                downloadCompiler(compilerVersion)
+                            }.fold(onSuccess = { it }, onFailure = {
+                                /**
+                                 * Try falling back on the version without the -EAP-SNAPSHOT suffix if the download
+                                 * for it fails - not all versions have a corresponding -EAP-SNAPSHOT version present
+                                 * in the snapshot repository.
+                                 */
+                                if (compilerVersion.endsWith(IntelliJPluginConstants.EAP_SUFFIX)) {
+                                    val nonEapVersion = compilerVersion.replace(
+                                        IntelliJPluginConstants.EAP_SUFFIX, ""
+                                    )
+                                    downloadCompiler(nonEapVersion)
+                                } else {
+                                    throw it
+                                }
                             })
                         } else {
                             warn(

--- a/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/IntelliJPlugin.kt
@@ -937,8 +937,8 @@ open class IntelliJPlugin : Plugin<Project> {
             }))
             cliVersion.convention(IntelliJPluginConstants.VERSION_LATEST)
             cliPath.convention(project.provider {
-                val resolvedCliVersion = SignPluginTask.resolveCliVersion(cliVersion.orNull)
-                val url = SignPluginTask.resolveCliUrl(resolvedCliVersion)
+                val resolvedCliVersion = resolveCliVersion(cliVersion.orNull)
+                val url = resolveCliUrl(resolvedCliVersion)
                 debug(context, "Using Marketplace ZIP Signer CLI in '$resolvedCliVersion' version")
 
                 dependenciesDownloader.downloadFromRepository(logCategory(), {

--- a/src/main/kotlin/org/jetbrains/intellij/IntelliJPluginConstants.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/IntelliJPluginConstants.kt
@@ -31,6 +31,15 @@ object IntelliJPluginConstants {
     const val ANNOTATIONS_DEPENDENCY_VERSION = "22.0.0"
     const val DEFAULT_IDEA_VERSION = "LATEST-EAP-SNAPSHOT"
 
+    const val RELEASE_SUFFIX_EAP = "-EAP-SNAPSHOT"
+    const val RELEASE_SUFFIX_EAP_CANDIDATE = "-EAP-CANDIDATE-SNAPSHOT"
+    const val RELEASE_SUFFIX_SNAPSHOT = "-SNAPSHOT"
+    const val RELEASE_SUFFIX_CUSTOM_SNAPSHOT = "-CUSTOM-SNAPSHOT"
+
+    const val RELEASE_TYPE_SNAPSHOTS = "snapshots"
+    const val RELEASE_TYPE_NIGHTLY = "nightly"
+    const val RELEASE_TYPE_RELEASES = "releases"
+
     const val MARKETPLACE_HOST = "https://plugins.jetbrains.com"
     const val IDEA_PRODUCTS_RELEASES_URL = "https://www.jetbrains.com/updates/updates.xml"
 //    const val ANDROID_STUDIO_PRODUCTS_RELEASES_URL = "https://dl.google.com/android/studio/patches/updates.xml"
@@ -45,6 +54,5 @@ object IntelliJPluginConstants {
     const val PLUGIN_PATH = "plugin.path"
     const val VERSION_LATEST = "latest"
     const val ANDROID_STUDIO_TYPE = "AI"
-    const val EAP_SUFFIX = "-EAP-SNAPSHOT"
 }
 

--- a/src/main/kotlin/org/jetbrains/intellij/IntelliJPluginConstants.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/IntelliJPluginConstants.kt
@@ -45,5 +45,6 @@ object IntelliJPluginConstants {
     const val PLUGIN_PATH = "plugin.path"
     const val VERSION_LATEST = "latest"
     const val ANDROID_STUDIO_TYPE = "AI"
+    const val EAP_SUFFIX = "-EAP-SNAPSHOT"
 }
 

--- a/src/main/kotlin/org/jetbrains/intellij/dependency/IdeaDependencyManager.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/dependency/IdeaDependencyManager.kt
@@ -11,6 +11,8 @@ import org.gradle.internal.os.OperatingSystem
 import org.gradle.kotlin.dsl.create
 import org.gradle.tooling.BuildException
 import org.jetbrains.intellij.IntelliJIvyDescriptorFileGenerator
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_SUFFIX_SNAPSHOT
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_TYPE_SNAPSHOTS
 import org.jetbrains.intellij.debug
 import org.jetbrains.intellij.ideBuildNumber
 import org.jetbrains.intellij.ideaDir
@@ -259,7 +261,7 @@ open class IdeaDependencyManager @Inject constructor(
                 "com.jetbrains.intellij.goland" to "goland"
             }
             type == "RD" -> {
-                if (sources && releaseType == "snapshots") {
+                if (sources && releaseType == RELEASE_TYPE_SNAPSHOTS) {
                     warn(context, "IDE sources are not available for Rider SNAPSHOTS")
                     hasSources = false
                 }
@@ -284,7 +286,7 @@ open class IdeaDependencyManager @Inject constructor(
             mavenRepository("$repositoryUrl/$releaseType")
         }).first().let {
             debug(context, "IDE zip: " + it.path)
-            unzipDependencyFile(getZipCacheDirectory(it, project, type), it, type, version.endsWith("-SNAPSHOT"))
+            unzipDependencyFile(getZipCacheDirectory(it, project, type), it, type, version.endsWith(RELEASE_SUFFIX_SNAPSHOT))
         }
 
         info(context, "IDE dependency cache directory: $classesDirectory")
@@ -374,7 +376,7 @@ open class IdeaDependencyManager @Inject constructor(
                     depFile.name.endsWith(".zip") -> {
                         val cacheDirectory = getZipCacheDirectory(depFile, project, "IC")
                         debug(context, "IDE extra dependency '$name': " + cacheDirectory.path)
-                        unzipDependencyFile(cacheDirectory, depFile, "IC", version.endsWith("-SNAPSHOT"))
+                        unzipDependencyFile(cacheDirectory, depFile, "IC", version.endsWith(RELEASE_SUFFIX_SNAPSHOT))
                     }
                     else -> {
                         debug(context, "IDE extra dependency '$name': " + depFile.path)

--- a/src/main/kotlin/org/jetbrains/intellij/tasks/SignPluginTask.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/tasks/SignPluginTask.kt
@@ -33,7 +33,9 @@ open class SignPluginTask @Inject constructor(
 ) : ConventionTask() {
 
     companion object {
-        private const val LATEST_RELEASE_URL = "https://github.com/JetBrains/marketplace-zip-signer/releases/latest"
+        private const val MARKETPLACE_ZIP_SIGNER_RELEASES_URL = "https://github.com/JetBrains/marketplace-zip-signer/releases"
+        private const val LATEST_RELEASE_URL = "$MARKETPLACE_ZIP_SIGNER_RELEASES_URL/latest"
+        private const val RELEASE_DOWNLOAD_URL = "$MARKETPLACE_ZIP_SIGNER_RELEASES_URL/download/%VERSION%/marketplace-zip-signer-cli.jar"
 
         /**
          * Resolves the latest version available of the Marketplace ZIP Signer CLI using GitHub API.
@@ -50,25 +52,6 @@ open class SignPluginTask @Inject constructor(
             } catch (e: IOException) {
                 throw GradleException("Cannot resolve the latest Marketplace ZIP Signer CLI version")
             }
-        }
-
-        /**
-         * Resolves Marketplace ZIP Signer CLI version.
-         * If set to {@link IntelliJPluginConstants#VERSION_LATEST}, there's request to {@link #METADATA_URL}
-         * performed for the latest available verifier version.
-         *
-         * @return Marketplace ZIP Signer CLI version
-         */
-        fun resolveCliVersion(version: String?) = version?.takeIf { it != IntelliJPluginConstants.VERSION_LATEST }
-            ?: resolveLatestVersion()
-
-        /**
-         * Resolves Marketplace ZIP Signer CLI download URL.
-         *
-         * @return Marketplace ZIP Signer CLI download URL
-         */
-        fun resolveCliUrl(version: String?) = resolveCliVersion(version).let {
-            "https://github.com/JetBrains/marketplace-zip-signer/releases/download/$it/marketplace-zip-signer-cli.jar"
         }
     }
 
@@ -305,5 +288,24 @@ open class SignPluginTask @Inject constructor(
         }
 
         return args
+    }
+
+    /**
+     * Resolves Marketplace ZIP Signer CLI version.
+     * If set to {@link IntelliJPluginConstants#VERSION_LATEST}, there's request to {@link #METADATA_URL}
+     * performed for the latest available verifier version.
+     *
+     * @return Marketplace ZIP Signer CLI version
+     */
+    internal fun resolveCliVersion(version: String?) = version?.takeIf { it != IntelliJPluginConstants.VERSION_LATEST }
+        ?: resolveLatestVersion()
+
+    /**
+     * Resolves Marketplace ZIP Signer CLI download URL.
+     *
+     * @return Marketplace ZIP Signer CLI download URL
+     */
+    internal fun resolveCliUrl(version: String?) = resolveCliVersion(version).let {
+        RELEASE_DOWNLOAD_URL.replace("%VERSION%", it)
     }
 }

--- a/src/main/kotlin/org/jetbrains/intellij/utils.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/utils.kt
@@ -30,6 +30,13 @@ import org.jdom2.Document
 import org.jdom2.JDOMException
 import org.jdom2.output.Format
 import org.jdom2.output.XMLOutputter
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_SUFFIX_CUSTOM_SNAPSHOT
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_SUFFIX_EAP_CANDIDATE
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_SUFFIX_EAP
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_TYPE_NIGHTLY
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_TYPE_RELEASES
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_TYPE_SNAPSHOTS
+import org.jetbrains.intellij.IntelliJPluginConstants.RELEASE_SUFFIX_SNAPSHOT
 import org.jetbrains.intellij.dependency.IdeaDependency
 import org.jetbrains.intellij.model.ProductInfo
 import org.xml.sax.SAXParseException
@@ -137,13 +144,13 @@ fun collectJars(directory: File, filter: Predicate<File>): Collection<File> = wh
 }
 
 fun releaseType(version: String) = when {
-    version.endsWith("-EAP-SNAPSHOT") ||
-        version.endsWith("-EAP-CANDIDATE-SNAPSHOT") ||
-        version.endsWith("-CUSTOM-SNAPSHOT") ||
+    version.endsWith(RELEASE_SUFFIX_EAP) ||
+        version.endsWith(RELEASE_SUFFIX_EAP_CANDIDATE) ||
+        version.endsWith(RELEASE_SUFFIX_CUSTOM_SNAPSHOT) ||
         version.matches(MAJOR_VERSION_PATTERN.toRegex())
-    -> "snapshots"
-    version.endsWith("-SNAPSHOT") -> "nightly"
-    else -> "releases"
+    -> RELEASE_TYPE_SNAPSHOTS
+    version.endsWith(RELEASE_SUFFIX_SNAPSHOT) -> RELEASE_TYPE_NIGHTLY
+    else -> RELEASE_TYPE_RELEASES
 }
 
 fun error(logCategory: String? = null, message: String, e: Throwable? = null) = log(LogLevel.ERROR, logCategory, message, e)

